### PR TITLE
drivers: UC81xx partial refresh profiles

### PR DESCRIPTION
**Pull request by andysan**
# Overview

All UC81xx devices support partial refresh by only scanning a subset (window) of the display. However, there is currently no way to specify different configuration overrides for different refresh types (full/partial). Having this capability makes it possible to implement fast (differential) partial refresh if the vendor provides suitable LUTs. This pull request addresses that by introducing different refresh profiles in the device tree. 

It consists of the following major changes:
* Minor cleanups
* drivers: uc81xx: Specify overrides in a DT child node
* drivers: uc81xx: Add a separate partial refresh profile
* Support for more types of overrides (LUTs/PLL/VDCs)
* drivers: uc81xx: Add workaround for chips without autocopy
* shields: Add a GDEW042T2 with fast partial refresh (fast partial refresh example)

# Add support for separate refresh profiles

Add support for separate partial and full refresh profiles. This makes it possible to specify different parameters for full and partial refreshes. The design is similar to the proposed design for the SSD16XX (https://github.com/zephyrproject-rtos/zephyr/pull/48163). However, unlike the SSD16xx, the driver always supports partial refresh even if a partial refresh profile hasn't been provided. This seems like the best option for this chip as it always supports scanning of a subset of the display.

Devices that need custom LUTs and voltages can specify them separately for the full and partial profiles. The controller will be reset when changing profiles which means that profiles always override the default reset values. This means that it is, for example, possible to use default values and LUTs from OTP for a full refresh and a custom profile for partial refreshes.

# Differences between UC8176 and UC8179

UC8176 and UC8179 seem to handle frame buffers differently. According to the specification, UC8179 is capable of automatically copying from the current frame buffer to the old frame buffer. This is usually not a problem for most full refresh profiles since they tend to not perform differential refresh. However, when performing a fast differential refresh, it's necessary to maintain both frame buffers correctly. The UC8176 requires manual updates of the old frame buffer. This is enabled if the `auto_copy` quirk is set to `false`.

# Testing

This has so far only been tested on a GDEW042T2. I don't own any boards using any other UltraChip controllers, so I would really appreciate it if someone with access to a GDEW075T7 could test this driver.

# Open questions

## Partial refresh specification in device tree

The current version of the patch assumes that a device that doesn't have a partial refresh node implements partial refresh with using the default configuration (the reset values) in the controller. This is different from the model in SSD16XX.

There are roughly possible approaches when no refresh profile has been specified in the device tree:

1.  Never perform any type of partial refresh. This is what the current patch for the SSD16xx does.
2. Treat it as an empty refresh profile in the device tree. I.e., use the controller's default configuration. Current implementation in this PR.
3. Use the settings from the full refresh profile, but only scan the refresh window. This is similar to the old behaviour.

I'm leaning towards updating the PR to use option 3 instead of option 2. Option 1 seems like a poor fit for this controller since it always supports a partial display refresh.

### CODE REVIEW

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay [hunk 1]
```diff
@@ -0,0 +1,95 @@
+/*
```
Comment (jfischer-no) [inline boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay:1]: You need additional changes to use it properly as a shield. There is also a PR that cleans it up https://github.com/zephyrproject-rtos/zephyr/pull/48791.

```diff
diff --git a/boards/shields/waveshare_epaper/Kconfig.defconfig b/boards/shields/waveshare_epaper/Kconfig.defconfig
index 10263e88bf2..b60e593e8cd 100644
--- a/boards/shields/waveshare_epaper/Kconfig.defconfig
+++ b/boards/shields/waveshare_epaper/Kconfig.defconfig
@@ -5,7 +5,8 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
-if SHIELD_WAVESHARE_EPAPER_GDEH029A1 || SHIELD_WAVESHARE_EPAPER_GDEH0213B1 || SHIELD_WAVESHARE_EPAPER_GDEH0213B72 || SHIELD_WAVESHARE_EPAPER_GDEW075T7 || SHIELD_WAVESHARE_EPAPER_GDEH0154A07 || SHIELD_WAVESHARE_EPAPER_GDEW042T2
+if SHIELD_WAVESHARE_EPAPER_GDEH029A1 || SHIELD_WAVESHARE_EPAPER_GDEH0213B1 || SHIELD_WAVESHARE_EPAPER_GDEH0213B72 || SHIELD_WAVESHARE_EPAPER_GDEW075T7 || SHIELD_WAVESHARE_EPAPER_GDEH0154A07 || SHIELD_WAVESHARE_EPAPER_GDEW042T2 || SHIELD_WAVESHARE_EPAPER_GDEW042T2_P
+
 
 
 if DISPLAY
diff --git a/boards/shields/waveshare_epaper/Kconfig.shield b/boards/shields/waveshare_epaper/Kconfig.shield
index 11b5567d7ed..863059e5784 100644
--- a/boards/shields/waveshare_epaper/Kconfig.shield
+++ b/boards/shields/waveshare_epaper/Kconfig.shield
@@ -18,3 +18,6 @@ config SHIELD_WAVESHARE_EPAPER_GDEH0154A07
 
 config SHIELD_WAVESHARE_EPAPER_GDEW042T2
        def_bool $(shields_list_contains,waveshare_epaper_gdew042t2)
+
+config SHIELD_WAVESHARE_EPAPER_GDEW042T2_P
+       def_bool $(shields_list_contains,waveshare_epaper_gdew042t2-p)
```

**Comment by andysan** [comment]
@jfischer-no Any thoughts on how a missing partial refresh profile should be treated? Is the current approach acceptable? Or should I just use the settings from the full refresh profile (if present) and just scan the partial window?

**Comment by jfischer-no** [comment]
>I would really appreciate it if someone with access to a GDEW075T7 could test this driver.

Tested, I do not see any regressions.

> @jfischer-no Any thoughts on how a missing partial refresh profile should be treated? Is the current approach acceptable? Or should I just use the settings from the full refresh profile (if present) and just scan the partial window?

+1  for "Use the settings from the full refresh profile, but only scan the refresh window. This is similar to the old behaviour."

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+};
+
+&arduino_spi {
+	/*
+	 * GoodDisplay GDEW042T2 with fast partial refresh. Based on
+	 * configuration from GoodDisplay's Arduino example.
+	 */
+	uc8176: uc8176@0 {
+		compatible = "gooddisplay,gdew042t2", "ultrachip,uc8176";
+		label = "UC8176";
```
[the remark this answers is not in the document]
Comment (andysan) [reply boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay:22]: Done.

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay (hunk 1 above)
Comment (andysan) [reply boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay:1]: Done.

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay (hunk 2 above)
Comment (galak) [inline boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay:22]: remove label properties.

**Review by galak — DISMISSED** [review]
remove `label` prop

File: boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay (hunk 2 above)
Comment (andysan) [reply boards/shields/waveshare_epaper/waveshare_epaper_gdew042t2-p.overlay:22]: Done

**Review by mbolivar-nordic — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**